Improve error reporting in rtc-to-net

bemasc opened the pull request:
This change attempts to improve the level of detail of error reporting without
revealing information that would allow a port scan of a private network.

Care is taken relative to https://github.com/uProxy/uproxy/issues/809

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy-networking/187)

<!-- Reviewable:end -->

bemasc committed:
Improve error reporting in rtc-to-net

This change attempts to improve the level of detail of error reporting without
revealing information that would allow a port scan of a private network.

trevj commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187)

---

Very nice. Kudos for anticipating out the port scanning trick!

---

<sup>**[src/rtc-to-net/rtc-to-net.ts, line 227 [r1]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187#-Jgmv75-r7JMtnsf0GOU)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/9ae3bf88d96fda011416639a9646689d8d92a0fd/src/rtc-to-net/rtc-to-net.ts#L227)):</sup>
It might be prettier to move this to the previous line...it'll fit!

---

<sup>**[src/rtc-to-net/rtc-to-net.ts, line 228 [r1]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187#-JgmxSuaSB3Pp4guOUtI)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/9ae3bf88d96fda011416639a9646689d8d92a0fd/src/rtc-to-net/rtc-to-net.ts#L228)):</sup>
It looks like `UNSUPPORTED_COMMAND` is one of (maybe the only?) `Reply` fields you don't have a test for. Maybe add a TODO?

---

<sup>**[src/rtc-to-net/rtc-to-net.ts, line 316 [r1]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187#-JgmwSpS38wWJ6W3bhgU)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/9ae3bf88d96fda011416639a9646689d8d92a0fd/src/rtc-to-net/rtc-to-net.ts#L316)):</sup>
Maybe it would be better to make `Response.endpoint` an optional field?

---

<sup>**[src/rtc-to-net/rtc-to-net.ts, line 334 [r1]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187#-Jgmxe98vedvRAYrPuM9)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/9ae3bf88d96fda011416639a9646689d8d92a0fd/src/rtc-to-net/rtc-to-net.ts#L334)):</sup>
So it's fine as is but I think you can improve this and `reportFailureToPeer_` by having them return, synchronously, a `Socks.Reply` value and restructuring `start()` appropriately.

What do you think? It would remove a lot of calls to `replyToPeer_` which are currently scattered across several methods.

---

<!-- Sent from Reviewable.io -->

bemasc committed:
Cleanup in response to review comments.

bemasc commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187)

---

<sup>**[src/rtc-to-net/rtc-to-net.ts, line 227 [r1]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187#-Jgmv75-r7JMtnsf0GOU)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/9ae3bf88d96fda011416639a9646689d8d92a0fd/src/rtc-to-net/rtc-to-net.ts#L227)):</sup>
Done.

---

<sup>**[src/rtc-to-net/rtc-to-net.ts, line 228 [r1]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187#-JgmxSuaSB3Pp4guOUtI)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/9ae3bf88d96fda011416639a9646689d8d92a0fd/src/rtc-to-net/rtc-to-net.ts#L228)):</sup>
Done.

---

<sup>**[src/rtc-to-net/rtc-to-net.ts, line 316 [r1]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187#-JgmwSpS38wWJ6W3bhgU)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/9ae3bf88d96fda011416639a9646689d8d92a0fd/src/rtc-to-net/rtc-to-net.ts#L316)):</sup>
Done.  Note that this creates a protocol break.

---

<sup>**[src/rtc-to-net/rtc-to-net.ts, line 334 [r1]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187#-Jgmxe98vedvRAYrPuM9)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/9ae3bf88d96fda011416639a9646689d8d92a0fd/src/rtc-to-net/rtc-to-net.ts#L334)):</sup>
Done.

---

<!-- Sent from Reviewable.io -->

trevj commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187)

---

:+1:

Yes!

---

<sup>**[src/rtc-to-net/rtc-to-net.ts, line 325 [r2]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187#-Jgr687bnzoabOTgtLEI)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/9afc8ded45e678aa4e879284bdc63ca93188ba48/src/rtc-to-net/rtc-to-net.ts#L325)):</sup>
Just FYI you could set this field inline...

---

<!-- Sent from Reviewable.io -->

bemasc commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187)

---

Argh.  When testing this I found that if the user attempts to reach a disallowed endpoint, it kills that RtcToNet session from then on, because it tries to send a duplicate error response on a data channel that has already been closed.  It needs a logic change to survive those events.

---

<!-- Sent from Reviewable.io -->

bemasc committed:
Fix double-error-response problem and add a unit test.

The unit test doesn't actually catch this exact problem, because
the exception is isolated inside the web worker where it is thrown,
but at least it should catch some related category of problems.

bemasc commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187)

---

I've added a fix for the issue I observed.

---

<sup>**[src/rtc-to-net/rtc-to-net.ts, line 325 [r2]](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187#-Jgr687bnzoabOTgtLEI)** ([raw file](https://github.com/uproxy/uproxy-networking/blob/9afc8ded45e678aa4e879284bdc63ca93188ba48/src/rtc-to-net/rtc-to-net.ts#L325)):</sup>
Done.

---

<!-- Sent from Reviewable.io -->

trevj commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-networking/187)

---

Still :+1:!

---

<!-- Sent from Reviewable.io -->
